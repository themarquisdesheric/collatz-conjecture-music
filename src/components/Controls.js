import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import Arrows from './Arrows';
import SelectInput from './SelectInput';

export default class Controls extends Component {
  static propTypes = {
    selected: PropTypes.string.isRequired,
    handleCollatz: PropTypes.func.isRequired,
    handleWave: PropTypes.func.isRequired,
    sequence: PropTypes.arrayOf(PropTypes.number).isRequired,
  }

  state = {
    startVal: 0
  };

  componentDidMount() {
    window.addEventListener('keydown', this.handleKeyDown);
    
    this.input.focus();
  }

  componentDidUpdate(_prevProps, prevState) {
    const { startVal } = this.state;

    if (prevState.startVal !== startVal) {
      this.props.handleCollatz(startVal);
      window.scrollTo(0, 0);
    }
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.handleKeyDown);
  }
    
  handleSubmit = (e) => {
    const startVal = Number(this.input.value);

    e.preventDefault();

    if (startVal < 2) {
      alert('You must enter a number greater than 1');
    } else {
      this.setState({ startVal });
    }
  }

  handleKeyDown = ({ key }) => {
    // prevent interfering with navigation arrows
    if (document.activeElement === this.input) return;
   
    if (key === 'ArrowLeft') this.handleDecrement();
    else if (key === 'ArrowRight') this.handleIncrement();
  }

  handleDecrement = () => {
    const { startVal } = this.state;

    if (startVal <= 2) {
      alert('You must enter a number greater than 1');
      return;
    } 

    this.setState({ startVal: startVal - 1 });
  }
  
  handleIncrement = () => {
    this.setState({ startVal: this.state.startVal + 1 });
  }

  render() {
    const { selected, handleWave, sequence } = this.props;

    return (
      <Fragment>
        {!!sequence.length && 
          <Arrows 
            handleIncrement={this.handleIncrement}
            handleDecrement={this.handleDecrement}
          />
        }
        
        <div className="controls-wrapper">
          <form onSubmit={this.handleSubmit}>
            <label>
              Start value
              <input
                type="number"
                // ! there is a bug here where the defaultValue doesn't update when using left/right arrows 
                // ! even though using sequence[0] outside of defaultValue works entirely as expected
                defaultValue={sequence[0] || 15}
                ref={node => this.input = node}
              />
            </label>
            <SelectInput 
              label="Wave type"
              selected={selected}
              onChange={handleWave}
            />
            <span className="button-wrapper">
              <button type="submit">
                â–· Play
              </button>
            </span>
          </form>
        </div>
      </Fragment>
    );
  }
}
