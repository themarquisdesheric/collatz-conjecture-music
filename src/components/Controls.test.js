import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Controls from './Controls';

describe('controls', () => {
  const props = {
    sequence: [],
    wave: 'sine',
    isPlaying: false,
    handleCollatz: jest.fn(),
    handleWave: jest.fn(),
    handleRepeat: jest.fn()
  }
  
  let wrapper = mount(<Controls {...props} />);
  
  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  
  it('should render playing spinner if isPlaying is true', () => {
    props.isPlaying = true;

    wrapper = mount(<Controls {...props} />);

    expect(wrapper.find('Loader').length).toBe(1);
  });
  
  it('should render Arrows if sequence.length is true', () => {
    props.sequence = [1];

    wrapper = mount(<Controls {...props} />);

    expect(wrapper.find('Arrows').length).toBe(1);
  });
});