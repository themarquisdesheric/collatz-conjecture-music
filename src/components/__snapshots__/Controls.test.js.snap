// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controls should match snapshot 1`] = `
<Controls
  handleCollatz={[MockFunction]}
  handleRepeat={[MockFunction]}
  handleWave={[MockFunction]}
  isPlaying={false}
  sequence={Array []}
  wave="sine"
>
  <div
    className="controls-wrapper"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Start value
        <input
          defaultValue={15}
          type="number"
        />
      </label>
      <SelectInput
        label="Wave type"
        onChange={[MockFunction]}
        wave="sine"
      >
        <label>
          Wave type
          <div>
            sine
          </div>
          <select
            onChange={[MockFunction]}
            value="sine"
          >
            <option
              key="sine"
              value="sine"
            >
              sine
            </option>
            <option
              key="sawtooth"
              value="sawtooth"
            >
              sawtooth
            </option>
            <option
              key="triangle"
              value="triangle"
            >
              triangle
            </option>
            <option
              key="square"
              value="square"
            >
              square
            </option>
          </select>
        </label>
      </SelectInput>
      <span
        className="button-wrapper"
      >
        <button
          type="submit"
        >
          â–· Play
        </button>
      </span>
    </form>
  </div>
</Controls>
`;
